### PR TITLE
#17768: Float32 support for Training mode in Batch Norm

### PR DESCRIPTION
**Pull request by VirdhatchaniKN**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/17768

### Problem description
To Provide Fp32 support for Training mode of BN

### What's changed
Support provided for fp32 data type for training mode of BN

### Checklist
- [x] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/runs/13224654999)
- [x] [Blackhole post-commit tests](https://github.com/tenstorrent/tt-metal/actions/runs/13224893567)
- [ ] [(Single-card) Tests for new models](https://github.com/tenstorrent/tt-metal/actions/runs/13224894613)
- [x] [(Single-card) Demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/13224895428) - passed as in main
- [x] [(Single-card) Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/runs/13224896084)
- [x] [(Single-card) Model perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/13224896401)

### CODE REVIEW

**Review by umadevimcw — APPROVED** [review]
Is sweep tests added to BN already? if not please add it

**Review by dmakoviichuk-tt — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/test_batch_norm.py [hunk 1]
```diff
@@ -13,6 +13,114 @@
...
+        running_mean=mean_data,
+        running_var=var_data,
+        weight=weight_data,
+        bias=bias_data,
+        training=training,
+        eps=eps,
+        momentum=momentum,
+    )
+    comp_pass = compare_results_batch_norm([tt_output], [torch_result])
+    if training:
```
Comment (sjameelTT) [inline tests/ttnn/unit_tests/operations/test_batch_norm.py:100]: comp_pass_1 and comp_pass_2 are unclear. Could you choose better variable names and explain what's going on here?

Comment (VirdhatchaniKN) [reply tests/ttnn/unit_tests/operations/test_batch_norm.py:100]: comp_pass_1 and comp_pass_2 is to compare PCC  for updated running stats. Now I've updated the names

**Review by sjameelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/running_statistics_sfpu_kernel.cpp [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+        if constexpr (old_running_mean_has_value) {
+            // 1 - momentum
+            cb_reserve_back(cb_tmp1, onetile);
+            sub_binary_tile_init();
+            tile_regs_acquire();
+            tile_regs_wait();
+            copy_tile_to_dst_init_short_with_dt(cb_momentum, cb_one);
+            for (uint32_t i = 0; i < onetile; ++i) {
+                copy_tile(cb_one, i, i * 2);
```
Comment (yugaoTT) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/running_statistics_sfpu_kernel.cpp:49]: I think this is leaving the space for second tile in dst? probably remove hardcoding of 2, since it might be more performant to copy more tiles into dst in the future

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/running_statistics_sfpu_kernel.cpp [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+            }
+            tile_regs_release();
+            cb_push_back(cb_tmp2, onetile);
+            cb_pop_front(cb_batch_var, onetile);
+
+            // cb_tmp1 * running stats --> (1 - momentum) * running stats
+            cb_wait_front(cb_tmp1, onetile);
+            cb_wait_front(cb_old_running_var, onetile);
+            cb_reserve_back(cb_tmp3, onetile);
+            mul_binary_tile_init();
```
Comment (yugaoTT) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/running_statistics_sfpu_kernel.cpp:175]: redundant init?

**Review by yugaoTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/running_statistics_sfpu_kernel.cpp (hunk 2 above)
Comment (KalaivaniMCW) [reply ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/running_statistics_sfpu_kernel.cpp:49]: @yugaoTT for binary SFPU operation, 
we are copying the `input_a` tiles to  DST[0], DST[2], DST[4].. and `input_b` tiles to  DST[1], DST[3], DST[5].. 
after binary operation is performed, the result is stored in `input_a's` tiles in DST  i.e. DST[0], DST[2], DST[4].. which we then pack to the output CB

**Comment by VirdhatchaniKN** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/13301503214 - passed

**Comment by VirdhatchaniKN** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/13301502230 - passed
